doc/parameter_server.md

```diff
@@ -0,0 +1,144 @@
+
+- Even when a single process can hold the global model, we might still shard it onto multiple parameter server instances, who share the both communication and optimization workload.
+
+There are several kinds of parameter to be handled separately:
+
+- Very big embedding table: Embedding table is a collection of <item id, embedding vector> pairs. There is also a mapping from item id to pserver id, so the corresonding embedding vector will be stored at the certain pserver
+
+- Big dense tensor: If the dense tensor parameter exceeds certain size, it will be sliced into several subtensors. The parameter name combining subtensor id will become a unique key, and the value is the subtensor(Please note that the slice tensor operation is zero-copy)
+
+- Small dense tensor: The small dense tensor parameter will be stored at certain pserver wholely
+
+
+The local model on workers and the global model on the parameter server(s) have the same size. In case that large embedding tables make the model size too big to fit in a single process's memory space, an intuitive solution is to save the model copies on an external storage service like Redis or Memcached.
+
+However, such a solution creates a complex workflow which introduces too many cross-node communications.
+
+We propose an alternative solution that doesn't rely on Redis or Memcachd. Instead, the global model, including the large embedding tables, is sharded across parameter servers, and the local model contains only part of the embedding table -- the small fraction that is required to compute recent minibatch(es).
+
+### Parameter Initialization
+
+
+Parameter initialization of a very big embedding table is lazy. For example, in online learning, there could be unkown item id in the training data. So, until worker send the unkown item id to pserver, will pserver initialize corresponding embedding vector and send back to worker. This is a `get_or_create` semantic.
+
+Other parameters could be initialized before training.
+
+
+## Workflow
+
+### Assumption
+
+We apply different solutions to different scenarios.
+
+
+| scenario| hardware | update stragety | communication  stragety |
+| :----  |:----  |:---|:---|
+| Recommendation  | CPU|  Async |  Parameter server |
+| CV | GPU | BSP| Allreduce|
+| NLP | GPU | BSP| Allreduce |
+
+
+Following is the assumption under recommendation scenario:
+
+- only support asynchronous SGD
+- only support worker failover, do not support pserver failover
+
+
+**Note**
+
+Allreduce failover and pserver failover will be discussed on other design docs.
+
+### Master
+
+- responsible for creating/deleting/scheduling pserver and worker
+- define parameter sharding strategy, and generate unique key for each parameter(or sharded parameter)
+- define the hash strategy from parameter key to pserver id 
+
+Workflow:
+
+
```

Comment:
Again, `Workflow:` here is not a valid subject. And these are not valid sentences.